# Make Dummies from Categorical variables

#### Imports I

In [2]:
import pandas as pd
import numpy as np

In [3]:
%store -r data
%store -r col_year_month
%store -r col_categorical
%store -r col_target
%store -r col_numeric
%store -r combined_list

### Just a reminder:
The DataFrame `data` includes the binarized `target` variable, has no NaNs, and is two columns more than `combined_list`, because two of the original categorical variables were redundant, `carrier_name` and `airport_name`.

In [4]:
data[combined_list].head()

year  month carrier airport  arr_flights  carrier_ct  weather_ct  nas_ct  \
0  2016      1      AA     DFW      11956.0      507.33       39.83  452.65   
1  2016      1      AA     DTW        588.0       34.71        4.71   19.77   
2  2016      1      AA     SEA        607.0       35.52        5.40   23.55   
3  2016      1      AA     JFK       1595.0      117.00       10.26  117.66   
4  2016      1      AA     SJC        327.0       23.67        1.47   14.50   

   security_ct  late_aircraft_ct  arr_cancelled  arr_diverted  arr_delay  \
0         4.72            529.47          201.0           9.0   106950.0   
1         0.32             38.49           13.0           2.0     5170.0   
2         0.00             27.52           12.0           2.0     4485.0   
3         0.83             89.25          137.0           6.0    23698.0   
4         0.06             19.29            0.0           0.0     2276.0   

   carrier_delay  weather_delay  nas_delay  security_delay  \
0        50027.0         2842.0    13913.0           167.0   
1         1754.0          312.0      655.0            16.0   
2         1857.0          198.0      715.0             0.0   
3         9681.0         1592.0     4371.0            18.0   
4          985.0           94.0      373.0             5.0   

   late_aircraft_delay  target  
0              40001.0       1  
1               2433.0       1  
2               1715.0       1  
3               8036.0       1  
4                819.0       1

### Display the columns to be transformed into dummies in green

In [5]:
# the `data2` variable holds the `combined_list` because when we color the redundant columns, the 
# programming that does the coloring disables other DataFrame methods and functions we will need for 
# further programming... need to do %store -r data; %store -r combined_list;
# then run this cell
data_for_dummies = data[combined_list].sample(5)
# function examines value's length, and returns red if its too long
def if_str_violet(val):
    val = val; color = 'darkviolet' if type(val) == str else 'black'; return 'color: %s' % color
# apply the function
red_dummies = data_for_dummies.style.applymap(if_str_violet)
red_dummies

In [6]:
%store data_for_dummies

Stored 'data_for_dummies' (DataFrame)


In [7]:
red_dummies_cheat = "just uncomment this" + """def if_str_red(val):
    val = val; color = 'red' if type(val) == str else 'black'; return 'color: %s' % color
red_dummies = data_for_dummies.style.applymap(if_str_red)
red_dummies"""
red_dummies_cheat

"just uncomment thisdef if_str_red(val):\n    val = val; color = 'red' if type(val) == str else 'black'; return 'color: %s' % color\nred_dummies = data_for_dummies.style.applymap(if_str_red)\nred_dummies"

In [8]:
%store red_dummies_cheat

Stored 'red_dummies_cheat' (str)


#### Use pandas to get dummies for `data.carrier`

In [9]:
# since AA is first without caling drop_first=True, it is not present for concatenation
# with the dataset used for supervised machine learning algorithms, however,
# we may use the commented out code in order to include AA
# carrier_dummies = pd.get_dummies(data.carrier, prefix='carrier', prefix_sep='_')

carrier_dummies = pd.get_dummies(data[combined_list].carrier, prefix='carrier', prefix_sep='_', drop_first=True)
carrier_dummies.columns

Index([u'carrier_AS', u'carrier_B6', u'carrier_DL', u'carrier_EV',
       u'carrier_F9', u'carrier_HA', u'carrier_NK', u'carrier_OO',
       u'carrier_UA', u'carrier_VX', u'carrier_WN'],
      dtype='object')

In [10]:
len(carrier_dummies.columns)

11

#### Use pandas to get dummies for `data.airport`

In [11]:
# since ABE is first without calling drop_first=True, it is not present for concatenation
# wiht the datset used for supervised machine learning algorithms, however,
# we may use the commented out code in order to include ABE
# airport_dummies = pd.get_dummies(data.airport, prefix='airport', prefix_sep ='_')

airport_dummies = pd.get_dummies(data[combined_list].airport, prefix='airport', prefix_sep ='_', drop_first=True)
airport_dummies.columns

Index([u'airport_ABI', u'airport_ABQ', u'airport_ABR', u'airport_ABY',
       u'airport_ACK', u'airport_ACT', u'airport_ACV', u'airport_ACY',
       u'airport_ADK', u'airport_ADQ',
       ...
       u'airport_TYR', u'airport_TYS', u'airport_UST', u'airport_VLD',
       u'airport_VPS', u'airport_WRG', u'airport_WYS', u'airport_XNA',
       u'airport_YAK', u'airport_YUM'],
      dtype='object', length=309)

Calling `len(airport_dummies.columns)` shows there are 309 columns created by the dummies process, since we dropped the first one as its n-1 features.

In [12]:
len(airport_dummies.columns)

309

Our dataset now includes 320 new features to add... this will change the computation expenditure

In [13]:
len(carrier_dummies.columns) + len(airport_dummies.columns)

320

Concatenate `data[combined_list]` with the carrier dummies

In [14]:
data_full_00 = pd.concat([data[combined_list], carrier_dummies], axis=1)
data_full_00.head()

year  month carrier airport  arr_flights  carrier_ct  weather_ct  nas_ct  \
0  2016      1      AA     DFW      11956.0      507.33       39.83  452.65   
1  2016      1      AA     DTW        588.0       34.71        4.71   19.77   
2  2016      1      AA     SEA        607.0       35.52        5.40   23.55   
3  2016      1      AA     JFK       1595.0      117.00       10.26  117.66   
4  2016      1      AA     SJC        327.0       23.67        1.47   14.50   

   security_ct  late_aircraft_ct     ...      carrier_B6  carrier_DL  \
0         4.72            529.47     ...               0           0   
1         0.32             38.49     ...               0           0   
2         0.00             27.52     ...               0           0   
3         0.83             89.25     ...               0           0   
4         0.06             19.29     ...               0           0   

   carrier_EV  carrier_F9  carrier_HA  carrier_NK  carrier_OO  carrier_UA  \
0           0           0           0           0           0           0   
1           0           0           0           0           0           0   
2           0           0           0           0           0           0   
3           0           0           0           0           0           0   
4           0           0           0           0           0           0   

   carrier_VX  carrier_WN  
0           0           0  
1           0           0  
2           0           0  
3           0           0  
4           0           0  

[5 rows x 30 columns]

##### Up to 30 columns with the first addition of dummies

In [15]:
data_full_00.tail()

year  month carrier airport  arr_flights  carrier_ct  weather_ct  \
11188  2016     11      WN     LGB        203.0        7.52        0.00   
11189  2016     11      WN     LIT        208.0       11.82        0.29   
11190  2016     11      WN     MAF        250.0       11.88        0.00   
11191  2016     11      WN     MCI       1984.0       85.57        3.34   
11192  2016     11      UA     SAV          1.0        0.00        0.00   

       nas_ct  security_ct  late_aircraft_ct     ...      carrier_B6  \
11188    1.52          0.0             16.96     ...               0   
11189    7.21          0.0             13.68     ...               0   
11190    6.35          0.0             19.77     ...               0   
11191   55.88          0.0            112.21     ...               0   
11192    0.00          0.0              0.00     ...               0   

       carrier_DL  carrier_EV  carrier_F9  carrier_HA  carrier_NK  carrier_OO  \
11188           0           0           0           0           0           0   
11189           0           0           0           0           0           0   
11190           0           0           0           0           0           0   
11191           0           0           0           0           0           0   
11192           0           0           0           0           0           0   

       carrier_UA  carrier_VX  carrier_WN  
11188           0           0           1  
11189           0           0           1  
11190           0           0           1  
11191           0           0           1  
11192           1           0           0  

[5 rows x 30 columns]

##### Note:
It's disturbing to see that a DataFrame we modified to have 11173 rows, shows 11192 after calling ` .tail()`...

however, examining the length or calling `.index` shows that data_full is the right length

In [16]:
data_full_00.index

Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            11183, 11184, 11185, 11186, 11187, 11188, 11189, 11190, 11191,
            11192],
           dtype='int64', length=11173)

##### Or we can check all the relevant lengths of DataFrame groups and see if they match up... so concatenation should work

In [17]:
# this throws no error
assert len(data) == len(data[combined_list]) == len(carrier_dummies) == len(data_full_00) == len(airport_dummies)

#### Re-use the `data_full` variable and 
Concatenate `data_full_00` with the airport dummies

In [18]:
data_full_01 = df = pd.concat([data_full_00, airport_dummies], axis=1)
data_full_01.head()

year  month carrier airport  arr_flights  carrier_ct  weather_ct  nas_ct  \
0  2016      1      AA     DFW      11956.0      507.33       39.83  452.65   
1  2016      1      AA     DTW        588.0       34.71        4.71   19.77   
2  2016      1      AA     SEA        607.0       35.52        5.40   23.55   
3  2016      1      AA     JFK       1595.0      117.00       10.26  117.66   
4  2016      1      AA     SJC        327.0       23.67        1.47   14.50   

   security_ct  late_aircraft_ct     ...       airport_TYR  airport_TYS  \
0         4.72            529.47     ...                 0            0   
1         0.32             38.49     ...                 0            0   
2         0.00             27.52     ...                 0            0   
3         0.83             89.25     ...                 0            0   
4         0.06             19.29     ...                 0            0   

   airport_UST  airport_VLD  airport_VPS  airport_WRG  airport_WYS  \
0            0            0            0            0            0   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   airport_XNA  airport_YAK  airport_YUM  
0            0            0            0  
1            0            0            0  
2            0            0            0  
3            0            0            0  
4            0            0            0  

[5 rows x 339 columns]

Looking at the `.head()` or calling `len(df)`,  we see that there are now 328 columns

In [19]:
len(data_full_01.columns)

339

And by calling `data_full.columns_0x[18]`, we know that `target` is still in there, so, for machine learning, we will have to separate it from `data_full`

In [20]:
data_full_01.columns[18]

'target'

In [21]:
%store data_full_01

Stored 'data_full_01' (DataFrame)


## Dummies for months

In [22]:
# since 1 is first month without calling drop_first=True, it is not present for concatenation
# wiht the datset used for supervised machine learning algorithms, however,
# we may use the commented out code in order to include 1 (January)
# months_dummies = pd.get_dummies(data[combined_list].month, prefix='months', prefix_sep ='_')
months_dummies = pd.get_dummies(data[combined_list].month, prefix='months', prefix_sep ='_', drop_first=True)
months_dummies.head()

months_2  months_3  months_4  months_5  months_6  months_7  months_8  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   months_9  months_10  months_11  
0         0          0          0  
1         0          0          0  
2         0          0          0  
3         0          0          0  
4         0          0          0

In [23]:
%store months_dummies

Stored 'months_dummies' (DataFrame)


In [24]:
data_full_02 = df = pd.concat([data_full_01, months_dummies], axis=1)
data_full_02.head()

year  month carrier airport  arr_flights  carrier_ct  weather_ct  nas_ct  \
0  2016      1      AA     DFW      11956.0      507.33       39.83  452.65   
1  2016      1      AA     DTW        588.0       34.71        4.71   19.77   
2  2016      1      AA     SEA        607.0       35.52        5.40   23.55   
3  2016      1      AA     JFK       1595.0      117.00       10.26  117.66   
4  2016      1      AA     SJC        327.0       23.67        1.47   14.50   

   security_ct  late_aircraft_ct    ...      months_2  months_3  months_4  \
0         4.72            529.47    ...             0         0         0   
1         0.32             38.49    ...             0         0         0   
2         0.00             27.52    ...             0         0         0   
3         0.83             89.25    ...             0         0         0   
4         0.06             19.29    ...             0         0         0   

   months_5  months_6  months_7  months_8  months_9  months_10  months_11  
0         0         0         0         0         0          0          0  
1         0         0         0         0         0          0          0  
2         0         0         0         0         0          0          0  
3         0         0         0         0         0          0          0  
4         0         0         0         0         0          0          0  

[5 rows x 349 columns]

## Drop the `carrier`, and `airport` columns
since we used them to create dummies, they are now redundant

In [25]:
data_full_02.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11173 entries, 0 to 11192
Data columns (total 349 columns):
year                   int64
month                  int64
carrier                object
airport                object
arr_flights            float64
carrier_ct             float64
weather_ct             float64
nas_ct                 float64
security_ct            float64
late_aircraft_ct       float64
arr_cancelled          float64
arr_diverted           float64
arr_delay              float64
carrier_delay          float64
weather_delay          float64
nas_delay              float64
security_delay         float64
late_aircraft_delay    float64
target                 int64
carrier_AS             uint8
carrier_B6             uint8
carrier_DL             uint8
carrier_EV             uint8
carrier_F9             uint8
carrier_HA             uint8
carrier_NK             uint8
carrier_OO             uint8
carrier_UA             uint8
carrier_VX             uint8
carrier_WN         

In [26]:
len(data_full_02.columns)

349

In [27]:
data_for_models = data_full_02.drop(['year','carrier', 'airport', 'month'], axis=1)

In [28]:
len(data_for_models.columns)

345

## Store the variables

In [29]:
%store data_for_models

Stored 'data_for_models' (DataFrame)
